### ISSUE TITLE
Mamba picks very old dependency without a real need

### ISSUE THREAD

**Issue by johanneskoester** [mamba-org/mamba#1659]
This is perhaps just a matter of too much freedom, but it really leads to an unintuitive result which in fact causes technical issues.

This environment
```
channels:
  - bioconda
  - conda-forge
dependencies:
  - bwa ==0.7.17
  - samtools =1.12
  - picard =2.25
```

makes mamba pick r-base 3.2.2 from the r-channel (part of defaults I guess). This package is so old, that it seems to be even from before conda 2.0, hence containing those short prefixes, so that it leads to issues when deploying in a long path.

When adding a constraint like `r-base >=4.0` you see that actually there is a working solution with recent versions of r-base which does not contradict any of those packages explicitly defined in the environment.

I might be wrong here, but my first thought is that we could really avoid lots of papercuts for people if the solver would strive for getting the most recent stuff if no constraints are given. And maybe more direct dependencies should be preferred there.

Also, I find it problematic that r-base is chosen from defaults, although there is a newer r-base in conda-forge, which has a higher priority than the defaults channel in this environment.

Another thought is that maybe one could also try to detect pre-conda 2.0 packages and exclude them from the search space if they are not absolutely required to solve the requested env?

**Comment by wolfv** [comment]
are you using strict channel priority?

**Comment by johanneskoester** [comment]
Have tried that, but it does not make a difference here.

**Comment by johanneskoester** [comment]
Ok, I just did the same as you did, but with `apptainer shell docker://condaforge/mambaforge` instead of building a new image. I get this:
```
$ mamba list -n x | grep r-base
r-base                    3.2.2                         0  
```

Could it be a matter of the location and CDN showing something different based on from where the request is made?

**Comment by jonashaag** [comment]
What CPU architecture does that use?

**Comment by wolfv** [comment]
The virtual packages might also play a role (e.g. `__glibc`)

**Comment by wolfv** [comment]
There is a – maybe – similar issue where mamba picks an old version because of the ongoing openssl 3.0.0 migration (https://github.com/mamba-org/mamba/issues/1581). My idea for that issue is to have an additonal "solver" loop where mamba tries to maximize the directly selected specs.

**Comment by johanneskoester** [comment]
In addition to just trying to go by version, what about trying to prioritize by some additional heuristic, namely:
If possible, prefer packages with a build string (i.e. containing automatic pinnings and conda >2.0) in contrast to packages coming from the pre-pinning era.

**Comment by johanneskoester** [comment]
> What CPU architecture does that use?

I believe apptainer usually uses the architecture of the underlying machine. In my case that is linux/amd64.

**Comment by johanneskoester** [comment]
This particular problem is solved via the latest mamba release and setting strict priorities (https://conda-forge.org/docs/user/tipsandtricks.html). However, in theory, the problem could also occur with purely packages from conda-forge or bioconda, in case a solution selects a very old package in order to avoid e.g. conflicts. I still think it might be a good idea to somehow protect against that, e.g. by first solving dependencies while only using packages with a build hash in the package name (indicating that a modern conda build has been used which automatically infers pinnings as much as possible). Only if no solution is found with that, mamba could also include non-build-hash packages.

**Comment by wolfv** [comment]
I think it is also often the case that _metadata_ is not correct and there has to be a balance between the two. If mamba selects an old package (but is allowed to do so by the metadata) is it really "at fault"?

But yeah, adding an additional solver loop that tries to resolve for most recent packages is not a bad idea.

**Comment by johanneskoester** [comment]
> I think it is also often the case that _metadata_ is not correct and there has to be a balance between the two. If mamba selects an old package (but is allowed to do so by the metadata) is it really "at fault"?

I completely agree that this is not mambas fault, and indeed the metadata is to blame. However, reality is that it is very unlikely that there will ever be a state where all metadata from all channels will be 100% correct (1. humans are not perfect, 2. this kind of maintenance needs a lot of time, 3. even if conda-forge and perhaps bioconda might manage that at some point, there are also other channels that people might want to use, with less sophisticated tooling around the builds). Hence, I think mamba should be still be able to find the right solution given that metadata is wrong.

Just using the most recent version of each package might fail though, because of e.g. incomplete migrations or pinning incompatibilities between channels. Since the build-hash based conda-build versions give you a higher likelihood of having correct pinnings, I still think that might be a good heuristic (e.g. one could check whether the build field in repodata contains only a number (old conda-build, ignore) or a build hash (new conda-build, preferable). Is there anything I miss about that, which renders it a bad idea (maybe the OCI plans?)?

**Comment by johanneskoester** [comment]
Any comment on this?

**Comment by kevlened** [comment]
While this isn't mamba's fault, I thought I'd add another example. In this case, `megahit`'s dependencies aren't strict enough. It simply requests `python`, rather than a version, leading `micromamba` to install `python` 1.6, and causing `megahit` to fail when running `megahit --help`.

Repro:
1. build the file below: `docker build --platform linux/amd64 -t mamba .`
2. then run it: `docker run --platform linux/amd64 -it --rm mamba megahit --help`

```bash
FROM mambaorg/micromamba:1.3.1

ARG MAMBA_DOCKERFILE_ACTIVATE=1

RUN micromamba config set use_lockfiles false
RUN micromamba config set channel_priority strict

RUN micromamba install --yes \
	--channel=conda-forge \
	--channel=bioconda \
	wget \
	samtools \
	megahit
```

`megahit --help` works if we set add `python=3.10.9` explicitly:
```bash
FROM mambaorg/micromamba:1.3.1

ARG MAMBA_DOCKERFILE_ACTIVATE=1

RUN micromamba config set use_lockfiles false
RUN micromamba config set channel_priority strict

RUN micromamba install --yes \
	--channel=conda-forge \
	--channel=bioconda \
	wget \
	samtools \
	megahit \
	"python=3.10.9"
```
Ideally, the latest version of python would be resolved automatically.
